### PR TITLE
feat(orb): automatic orb releases

### LINKED ISSUE #12: Automatic orb releases
CircleCI orbs need to be released manually at the moment, using an org admin's token. It would be great to piggy-back on semantic-release and release a new patch, minor or major version of an orb based on the corresponding commit message.

### PR DESCRIPTION
**Pull request by rofe**
Fix #12

### CODE REVIEW

**Comment by rofe** [comment]
@dominique-pfister would be great to get a bash (script) review from you :)

**Comment by rofe** [comment]
what it's supposed to do:
1. dry-run semantic-release to analyze commits without triggering an npm release. if one of the commit messages suggests a patch, minor or major release, then...
2. go though orbs and compare local source with published source to see if the change was orb related. if there is a diff, then...
3. publish a new orb version using the previously detected version segment (patch, minor or major)

File: .circleci/config.yml [hunk 1]
```diff
@@ -66,6 +66,37 @@ jobs:
           name: Check for Updated Package name
           command: node -e "if (require('./package.json').name=='@adobe/helix-library') { process.exit(1) };"
 
+      - run:
+          name: Semantic Orb Release
+          command: |
+              # analyze commits to find out if any orbs need to be updated
+              npm run semantic-release --dry-run --analyze-commits | tee /tmp/commit-analysis.log
```
Comment (dominique-pfister) [inline .circleci/config.yml]: Will this script run concurrently on the same machine and filesystem? If yes, you might want to use temporary files, as shown here: https://www.cyberciti.biz/tips/shell-scripting-bash-how-to-create-empty-temporary-file-quickly.html

File: .circleci/config.yml [hunk 2]
```diff
@@ -66,6 +66,37 @@ jobs:
...
+              segment=$(cat /tmp/commit-analysis.log | sed '/^.* complete: */!d; s///;q' | sed 's/^\(.*\) release/\1/')
+              if [[ "${segment}" =~ ^(major|minor|patch)$ ]]; then
+                echo "Commit analysis suggests ${segment} release, identifying affected orbs"
+                for path in .circleci/orbs/*
+                do
```
Comment (dominique-pfister) [inline .circleci/config.yml]: For consistency in this script, you might use:
```
for path in .circleci/orbs/* ; do
```
(one line)

File: .circleci/config.yml [hunk 3]
```diff
@@ -66,6 +66,37 @@ jobs:
...
+                  if [[ "${path}" =~ .*/README.md$ ]]; then
+                    continue
+                  fi
+                  echo "Orb ${path}"
+                  orb=$(basename ${path})
+                  # run circleci first to check for updates
+                  circleci > /dev/null
+                  # compare published orb to local source
+                  circleci orb source adobe/${orb} > /tmp/${orb}-pub.yml
```
Comment (dominique-pfister) [inline .circleci/config.yml]: Same as above: might clash in concurrently running invocations

File: .circleci/config.yml [hunk 4]
```diff
@@ -66,6 +66,37 @@ jobs:
...
+              # analyze commits to find out if any orbs need to be updated
+              npm run semantic-release --dry-run --analyze-commits | tee /tmp/commit-analysis.log
+              segment=$(cat /tmp/commit-analysis.log | sed '/^.* complete: */!d; s///;q' | sed 's/^\(.*\) release/\1/')
+              if [[ "${segment}" =~ ^(major|minor|patch)$ ]]; then
+                echo "Commit analysis suggests ${segment} release, identifying affected orbs"
+                for path in .circleci/orbs/*
+                do
+                  if [[ "${path}" =~ .*/README.md$ ]]; then
+                    continue
+                  fi
```
Comment (dominique-pfister) [inline .circleci/config.yml]: One line `if` statements can be replaced with `condition && then-statement`, e.g.:
```
[[ "${path}" =~ .*/README.md$ ]] && continue
```

**Comment by rofe** [comment]
Reimplemented the logic in JS

File: src/orb-release/cli.js [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+      required: false,
+    })
+    .option('orb-src', {
+      type: 'string',
+      alias: 'orbSrc',
+      describe: 'The name of the orb source file',
+      default: 'orb.yml',
+      required: false,
+    })
+    .option('changelog', {
```
Comment (kptdobe) [inline src/orb-release/cli.js]: Why `changeLog` and not `change-log` ? (same pattern than other options)

File: src/orb-release/cli.js [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+  'patch',
+  'minor',
+  'major',
+];
+const logger = console; // executed in CircleCI runtime
+let config = {};
+
+function getReleaseDate() {
+  if (fs.existsSync(config.changelog)) {
+    const cl = fs.readFileSync(config.changelog).toString('utf-8');
```
Comment (kptdobe) [inline src/orb-release/cli.js]: Filesystem sync operations should be avoided. You should rather do `await fs.exists(...)` and `const cl = await fs.readFile(...)` (+ cascading async functions).

File: src/orb-release/cli.js [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+function diffOrb(orb) {
+  logger.log('Checking orb %s', orb.name);
+  let ret = false;
+  const pubFile = `${orb.tmp}/src.yml`;
+  shell.exec(`~/circleci orb source ${config.namespace}/${orb.name} > "${pubFile}"`);
+  if (shell.error()) {
+    return ret;
+  }
+  const pubSource = fs.readFileSync(pubFile).toString('utf-8');
+  const locSource = `${fs.readFileSync(orb.src).toString('utf-8')}\n`;
```
Comment (kptdobe) [inline src/orb-release/cli.js]: Same here, no sync.

File: test/testOrbRelease.js [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+    logger.error.restore();
+    // add env variables back
+    process.env.CIRCLECI_CLI_TOKEN = originalAuth;
+  });
+
+  it('refuses to run without required arguments', async () => {
+    const output = await runShell({});
+    assert.notEqual(output.code, 0, `expected exit code != 0, but got ${output.code}`);
+    assert.ok(/Missing required argument: token/.test(output.stderr), 'expected missing required arguments');
+  }).timeout(10000);
```
Comment (kptdobe) [inline test/testOrbRelease.js]: You can avoid all the `timeout(10000)` after each `it` by doing the `timeout(...)` call in the `before`. See https://github.com/adobe/helix-cli/blob/master/test/testDeployCmd.js#L39

File: test/testStatuspage.js [hunk 9]
```diff
@@ -111,7 +111,7 @@ describe('Testing statuspage', () => {
     const output = await runShell();
     assert.notEqual(output.code, 0, `expected exit code != 0, but got ${output.code}`);
     assert.ok(/Not enough non-option arguments/.test(output.stderr), 'expected help output');
-  });
+  }).timeout(5000);
```
Comment (kptdobe) [inline test/testStatuspage.js]: Same with `timeout`, see previous comment.

File: src/orb-release/cli.js (hunk 5 above)
Comment (rofe) [reply src/orb-release/cli.js]: because the file is typically called `CHANGELOG.md`

File: test/testOrbRelease.js (hunk 8 above)
Comment (rofe) [reply test/testOrbRelease.js]: Cool, thanks.

File: src/orb-release/cli.js (hunk 6 above)
Comment (rofe) [reply src/orb-release/cli.js]: Ok. But why?

Comment (kptdobe) [reply src/orb-release/cli.js]: Sync operations will block the main thread from performing any other operations. They should be avoided in general.

Comment (rofe) [reply src/orb-release/cli.js]: Well, in this case there is nothing else which needs to be going on 😏

Comment (rofe) [reply src/orb-release/cli.js]: I looked into using async `fs` functions and quickly ran into several problems... mainly related to the fact that the async API methods usually don't return anything, so `await`ing them won't work... I'd have to refactor a lot of code to callback-based, which in this case - and in my opinion - is not really worth the effort: this code will be executed synchronously and only once per circleci job.

Duly noted, though :)

Comment (kptdobe) [reply src/orb-release/cli.js]: https://nodejs.org/api/fs.html#fs_fspromises_readfile_path_options is your friend ;)

Comment (kptdobe) [reply src/orb-release/cli.js]: We do that everywhere ;)
https://github.com/adobe/helix-cli/search?q=readFile%28&unscoped_q=readFile%28

**Pull request merged**